baseplate/context/sqlalchemy.py

```diff
@@ -3,14 +3,88 @@
 from __future__ import print_function
 from __future__ import unicode_literals
 
-from sqlalchemy import event
+from sqlalchemy import create_engine, event
+from sqlalchemy.engine.url import URL
 from sqlalchemy.orm import Session
 
-from ..context import ContextFactory
-from ..core import (
-    ServerSpan,
-    SpanObserver,
-)
+from baseplate import config
+from baseplate.context import ContextFactory
+from baseplate.core import ServerSpan, SpanObserver
+
+
+def engine_from_config(app_config, secrets=None, prefix="database.", **kwargs):
+    """Make an Engine from a configuration dictionary.
+
+    The keys useful to :py:func:`engine_from_config` should be prefixed, e.g.
+    ``database.drivername``, ``database.username``, etc. The ``prefix`` argument
+    specifies the prefix used to filter keys.  Each key is mapped to a
+    corresponding keyword argument on the :py:class:`sqlalchemy.engine.url.URL`
+    constructor.  The :py:class:`sqlalchemy.engine.url.URL` object will then be
+    passed to :py:func:`sqlalchemy.create_engine`.  Any additional keyword arguments
+    that are passed in will override the values set in ``app_config`` that are
+    passed to the :py:class:`sqlalchemy.engine.url.URL` constructor.
+
+    Supported keys:
+
+    * ``url`` (optional): the connection URL to the database, if this is supplied, it
+        will be the only value used to connect to the database, all others will be
+        ignored.
+    * ``drivername`` (optional):  the name of the database backend. This name will
+        correspond to a module in sqlalchemy/databases or a third party plug-in e.g.
+        "postgresql".  Note that this is only optional if ``url`` is set, otherwise
+        it will be required.
+    * ``username`` (optional): the username used to connect to the database.
+    * ``password_secret`` (optional): the key used to retrieve the password from
+        ``secrets`` as a :py:class:`baseplate.secrets.SimpleSecret`.
+    * ``host`` (optional): the name of the database host.
+    * ``port`` (optional): the port number that the database is listening on on
+    the host.
+    * ``database`` (optional): the database name.
+    * ``query`` (optional): a dictionary of options to be passed to the dialect
+        and/or the DBAPI upon connect.
+
+    """
+    assert prefix.endswith(".")
+    config_prefix = prefix[:-1]
+    cfg = config.parse_config(app_config, {
+        config_prefix: {
+            "url": config.Optional(config.String),
+            "drivername": config.Optional(config.String),
+            "username": config.Optional(config.String),
+            "password_secret": config.Optional(config.String),
```

Comment:
I think we probably want to have username be in here too. If the credentials are being rotated, that'd generally be done by changing user to a new one with a new password, rolling everyone off the old one, then killing the old user. Right?

Reply:
(do we want to add a "credentials" secret type?)

Reply:
🤔

Reply:
https://github.com/reddit/baseplate.py/pull/267